docs/proposals/20230407-flexible-managed-k8s-endpoints.md

```diff
@@ -0,0 +1,519 @@
+
+#### Core Cluster API Controllers changes
+
+- All the controllers working with ClusterClass objects must take into account that the `infrastructure.Ref` field could be omitted; most notably:
+  - The ClusterClass controller must ignore nil `infrastructure.Ref` fields while adding owner references to all the objects referenced by a ClusterClass.
+  - The Topology controller must skip the generation of the `<Infra>Cluster` objects when the `infrastructure.Ref` field in a ClusterClass is empty.
+
+- All the controllers working with Cluster objects must take into account that the `infrastructureRef` field could be omitted; most notably:
+  - The Cluster controller must use skip reconciling this external reference when the `infrastructureRef` is missing; also, the `status.InfrastructureReady` field must be automatically set to true in this case.
+
+- The Cluster controller must reconcile the new `ClusterEndpoint` CR. Please note that:
+  - The value from the `ClusterEndpoint` CRD must surface on the `spec.ControlPlaneEndpoint` field on the `Cluster` object.
+  - If both are present, the value from the `ClusterEndpoint` CRD must take precedence on the value from `<Infra>Cluster` objects still using the `spec.controlPlaneEndpoint`.
+
+- The Cluster controller must implement the temporary machinery to migrate to the new CRD existing Clusters and to deal with `<Infra>Cluster` objects still using the `spec.controlPlaneEndpoint` field as a way to communicate the ClusterAddress to "core" Cluster API controllers:
+  - If there is the `spec.ControlPlaneEndpoint` on the `Cluster` object but not a corresponding `ClusterEndpoint` CR, the CR must be created.
+
+#### Provider controller changes
+
+- All the `<Infra>Cluster` controller who are responsible to create a control plane endpoint
+  - As soon as the `spec.controlPlaneEndpoint` field in the `<Infra>Cluster` object will removed, the `<Infra>Cluster` controller must instead create a `ClusterEndpoint` CR to communicate the control plane endpoint to the Cluster API core controllers
+    - NOTE: technically it is possible to start creating the `ClusterEndpoint` CR *before* the removal of the `spec.controlPlaneEndpoint` field, because the new CR will take precedence on the value read from the field, but this is up to the infra provider maintainers.
+  - The `ClusterEndpoint` CR must have an owner reference to the `<Infra>Cluster` object from which it is originated.
+
+- All the `ControlPlane Provider` controller who are responsible to create a control plane endpoint
+  - Must stop to wait for the `spec.ControlPlaneEndpoint` field on the `Cluster` object to be set before starting to provision the control plane.
+  - As soon as the control plane endpoint is available, the controller must create a `ClusterEndpoint` CR to communicate the control plane endpoint to the Cluster API core controllers
```

Comment:
Have we considered the possibility that the user will create this CR? What should the behavior be if the CR already exists?

Reply:
> Have we considered the possibility that the user will create this CR?

This probably has to be a supported use case right? Given we position this CRD as a way to replace the controlPlaneEndpoint fields in InfraCluster CRDs and in InfraCluster users today can use this field to set a controlPlaneEndpoint